lib/tree.rb

```diff
@@ -16,40 +16,101 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(log n)
+  # Space Complexity: O(1)
   def add(key, value)
-    raise NotImplementedError
+    @root = add_helper(@root, key, value)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def add_helper(root, key, value)
+    if root.nil?
+      root = TreeNode.new(key, value)
+      return root
+    end
+
+    if root.key > key
+      root.left = add_helper(root.left, key, value)
+    else
+      root.right = add_helper(root.right, key, value)
+    end
+    
+    return root
+  end
+
+  # Time Complexity: O(log n)
+  # Space Complexity: O(log n)
   def find(key)
-    raise NotImplementedError
+    return find_helper(@root, key)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def find_helper(node, key)
+    return nil if node.nil?
+
+    if node.key == key
+      return node.value
+    elsif node.key < key
+      return find_helper(node.right, key)
+    else
+      return find_helper(node.left, key)
+    end
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def inorder
-    raise NotImplementedError
+    result = []
+    inorder_helper(@root, result)
+    return result
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def inorder_helper(node, result)
+    return if node.nil?
+  
+    inorder_helper(node.left, result)
+    result.push({:key => node.key, :value => node.value})
+    inorder_helper(node.right, result)
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def preorder
-    raise NotImplementedError
+    result = []
+    preorder_helper(@root, result)
+    return result
   end
+  def preorder_helper(node, result)
+    return if node.nil?
 
-  # Time Complexity: 
-  # Space Complexity: 
+    result.push({:key => node.key, :value => node.value})
+    preorder_helper(node.left, result)
+    preorder_helper(node.right, result)
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def postorder
-    raise NotImplementedError
+    result = []
+    postorder_helper(@root, result)
+    return result
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def postorder_helper(node, result)
+    return if node.nil?
+
+    postorder_helper(node.left, result)
+    postorder_helper(node.right, result)
+    result.push({:key => node.key, :value => node.value})
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def height
```

Comment:
The space complexity would be O(h), where h is the height of the tree, due to the call stack.